File: java/src/main/java/com/dsa/examples/sorting/InsertionSort.java
// Sorting — Insertion Sort

package com.dsa.examples.sorting;

/**
 * Insertion sort — efficient on nearly sorted input (O(n) best case).
 * Time: O(n²) worst/average, O(n) best; Space: O(1).
 */
public final class InsertionSort {

    private InsertionSort() {
    }

    /**
     * Sorts {@code arr} in ascending order in place.
     * Time: O(n²) worst; O(n) best when nearly sorted; Space: O(1).
     */
    public static void sort(int[] arr) {
        if (arr == null || arr.length < 2) {
            return;
        }
        for (int i = 1; i < arr.length; i++) {
            int key = arr[i];
            int j = i - 1;
            while (j >= 0 && arr[j] > key) {
                arr[j + 1] = arr[j];
                j--;
            }
            arr[j + 1] = key;
        }
    }
}

File: python/dsa/sorting/insertion_sort.py
# Sorting — Insertion Sort

"""Insertion sort — efficient on nearly sorted input (O(n) best case).

Time: O(n²) worst/average, O(n) best; Space: O(1).
"""


def sort(arr: list[int] | None) -> None:
    """Sorts arr in ascending order in place.

    Time: O(n²) worst; O(n) best when nearly sorted; Space: O(1).
    """
    if arr is None or len(arr) < 2:
        return
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
